scalding-core/src/main/scala/com/twitter/scalding/Mode.scala

```diff
@@ -63,72 +75,258 @@ object Mode {
+
+  /** Create a write-only tap which simply discards the results */
+  def createNullTap: Tap[_, _, _]
+
+  // access this through the DistributedCacheFile object
+  private[scalding] def addCachedFile(file: UncachedFile): CachedFile
+
+  /** Read the content of a file into a string, UTF-8 encoding is assumed */
+  def readFromFile(filename: String): String
+  /** Read the content of a file into a string, UTF-8 encoding is assumed */
+  def writeToFile(filename: String, text: String): Unit
+}
+
+object StorageMode {
+  trait TemporarySource[T] {
+    def sink(conf: Config): TypedSink[T]
+    def downstreamPipe(conf: Config): TypedPipe[T]
+  }
+}
+
+trait Mode extends java.io.Serializable {
+  def storageMode: StorageMode
+  def executionMode: ExecutionMode
+
+  def strictSources: Boolean
+  def name: String
+
+  // Override this if you define new testing modes
+  def isTesting: Boolean = false
+
+  /**
+   * Using a new FlowProcess, which is only suitable for reading outside
+   * of a map/reduce job, open a given tap and return the TupleEntryIterator
+   */
+  final def openForRead(config: Config, tap: Tap[_, _, _]): TupleEntryIterator = executionMode.openForRead(config, tap)
+
+  /** Create a new FlowConnector for this cascading planner */
+  protected // will probably have to end up just "@deprecated"
+  final def newFlowConnector(config: Config): FlowConnector = newFlowConnector(config)
+
+  /** Create a new Flow for this cascading planner */
+  def newFlow(config: Config, flowDef: FlowDef): Flow[_] = executionMode.newFlow(config, flowDef)
+
+  /**
+   * @return true if the file exists on the current filesystem.
+   */
+  final def fileExists(filename: String): Boolean = storageMode.fileExists(filename)
+
   @deprecated("A Config is needed, especially if any kryo serialization has been used", "0.12.0")
   final def openForRead(tap: Tap[_, _, _]): TupleEntryIterator =
     openForRead(Config.defaultFrom(this), tap)
+}
 
-  // Returns true if the file exists on the current filesystem.
-  def fileExists(filename: String): Boolean
-  /** Create a new FlowConnector for this cascading planner */
-  def newFlowConnector(props: Config): FlowConnector
+/**
+ * Any Mode that runs using local resources only
+ */
+trait LocalMode extends Mode {
+}
+
+/**
+ * Any Mode that runs over a cluster (as opposed to Local) is a ClusterMode.
+ */
+trait ClusterMode extends Mode {
 }
 
-trait HadoopMode extends Mode {
+/**
+ * Any ClusterMode whose main configuration is org.apache.hadoop.conf.Configuration is a HadoopFamilyMode
+ */
+trait HadoopFamilyMode extends ClusterMode {
+  def jobConf: Configuration
+}
+/**
+ * The "HadoopMode" is actually an alias for "a mode running on a fabric that ultimately runs using an execution
+ * engine compatible with some of the Hadoop technology stack (may or may not include Hadoop 1.x, YARN, etc.)
+ */
+trait HadoopExecutionModeBase[ConfigType <: Configuration]
```

Comment:
maybe we actually want an abstract type here:

```scala
trait HadoopExecutionModeBase {
  type ConfigType <: Configuration
}
```
since then we can refer to it: `mode.ConfigType`. We might do this to put more of the cascading types and not use `_` so much.